mne/time_frequency/csd.py

```diff
@@ -1346,5 +1347,46 @@ def _csd_morlet(data, sfreq, wavelets, nfft, tslice=None, use_fft=True,
 
     # Scaling by sampling frequency for compatibility with Matlab
     csds /= sfreq
-
     return csds
+
+
+@verbose
+def compute_csd(epochs_tfr, verbose=None, **kwargs):
+    """Compute covariance matrices across frequencies for TFR epochs.
+
+    Parameters
+    ----------
+    epochs_tfr : EpochsTFR
+        The time-frequency resolved epochs over which to compute the
+        covariance.
+    %(verbose)s
+    **kwargs : dict
+        Arguments to pass to :func:`mne.compute_covariance`.
+
+    Returns
+    -------
+    res : instance of CrossSpectralDensity
+        Cross-spectral density restricted to selected channels.
+    """
+    _validate_type(epochs_tfr, EpochsTFR)
+    n_channels, n_freqs = len(epochs_tfr.ch_names), epochs_tfr.freqs.size
+    data = np.zeros((n_channels * (n_channels + 1) // 2, n_freqs),
+                    dtype=np.complex128)
+    # iterate over epochs
+    for idx, epochs_data in enumerate(epochs_tfr.data):
+        epochs_data_conj = np.conj(epochs_data)
+        csds = np.vstack([np.mean(epochs_data[[i]] * epochs_data_conj[i:],
+                                  axis=2) for i in range(n_channels)])
```

Comment:
to avoid the for loop I would compute the full CSD matrix and use tri_indices to only store the upper part.

Reply:
I tried to refactor this but it's a tricky einsum because you need to preserve the frequency dimension but average across the time dimension so you could loop over frequencies to do that but it's pretty much just as inelegant. Maybe someone else could puzzle it out but I think this is a lot harder than it looks.

Reply:
Done in 25e3c82be9